celery/backends/base.py

```diff
@@ -170,6 +170,13 @@ def mark_as_failure(self, task_id, exc,
         if request:
             if request.chord:
                 self.on_chord_part_return(request, state, exc)
+            elif request.chain and 'chord' in request.chain[-1]['options']:
```

Comment:
does this solve the reported issue?

Reply:
Yes. The problem is that this snippet of code is executed for the task in the chain that fails, but only the final task in the chain passes the `if request.chord` check, so the chord only gets notified if it's the final task in the chain that fails. Adding this check solves the problem as bluntly as possible: by simply checking for a chord associated with the last element in the chain and updating it if necessary. Very much unsure if that is the right way forward. 

It looks like it also breaks a couple of other tests. I think fixing those tests would just be a matter of changing this line to:
```python
elif getattr(request, 'chain', None) and 'chord' in request.chain[-1]['options']:
```
but that just further reinforces my idea that this is maybe not the right way/place to be fixing this problem.

Reply:
keep digging with it please